### PR TITLE
Feature/core 746 contacts component

### PR DESCRIPTION
**Pull request by datrinh**
Implemented contact asset overview page

Favorite functionality extracted into own subtask:
https://evannetwork.atlassian.net/browse/CORE-789

### CODE REVIEW

File: dapps/assets/src/components/Assets.ts [hunk 1]
```diff
@@ -30,6 +30,19 @@ declare module 'vue/types/vue' {
...
+  interface Dapp {
+    baseHash: string;
+    baseUrl: string;
+    contractAddress: string;
+    domainName: string;
+    ens: string;
+    fullUrl: string;
+    rootEns: string;
   }
 }
```
Comment (Tschuck) [inline dapps/assets/src/components/Assets.ts:47]: This is nice! Tried this within a general shim file within the route `/vue/ts-shim.d.ts` folder, but the linter will not load this correctly. Would be nice to have it global, so it will work for every dapp automatically. Do you have a quick idea, how we can solve it globally?

File: dapps/assets/src/components/contacts/AddContact.ts [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+  msgTitle = this.$t('_assets.contacts.subject-prefill');
+
+  async created() {
+    this.initState();
+    const runtime = this.getRuntime();
+    this.contactService = new ContactsService(runtime);
+    this.addressbook = await runtime.profile.getAddressBook();
+  }
+
+  private initState() {
```
Comment (Tschuck) [inline dapps/assets/src/components/contacts/AddContact.ts:54]: `initState` function can be removed by setting initial values within the class head.

File: dapps/assets/src/components/contacts/AddContact.ts [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+      this.closePanel();
+
+      this.$emit('contact-added');
+    }
+  }
+
+  handleIdOrEmailChange(value: string) {
+    this.idOrEmailValidation = this.validateIdOrEmail(value);
+    console.log(this.idOrEmailValidation);
```
Comment (Tschuck) [inline dapps/assets/src/components/contacts/AddContact.ts]:
```suggestion
```

File: dapps/assets/src/components/contacts/AddContact.ts [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+import { EvanComponent, EvanForm } from '@evan.network/ui-vue-core';
+import { ContactsService } from './ContactsService';
+
+@Component
+export default class AddContactComponent extends mixins(EvanComponent) {
+  contactService: ContactsService;
+
+  addressbook;
+
+  idOrEmailValidation = '';
```
Comment (Tschuck) [inline dapps/assets/src/components/contacts/AddContact.ts]: Would suggest to something like `idOrEmailError` or `invalidIdOrEmail`, will be better understandable.

File: dapps/assets/src/components/contacts/AddContact.ts [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+        this.email = null;
+        return '';
+      }
+    } else if (EvanForm.validateEmail(value)) {
+      this.emailInvite = true;
+      this.email = value;
+      this.accountId = null;
+      return '';
+    } else {
+      return '_assets.contacts.error-id-or-email';
```
Comment (Tschuck) [inline dapps/assets/src/components/contacts/AddContact.ts:119]: String is missing in translations `error-id-or-email `.

File: dapps/assets/src/components/contacts/AddContact.vue [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+      alignment="right"
+      type="default"
+      class="light"
+      :showBackdrop="true"
+      :hideCloseButton="true"
+      :title="$t('_assets.contacts.add-contact-title')"
+      @hide="initState"
+    >
+      <p>{{ '_assets.contacts.add-contact-desc' | translate }}</p>
+        <evan-form-control-input
```
Comment (Tschuck) [inline dapps/assets/src/components/contacts/AddContact.vue]: Wrong indent. Please check also the next lines.
```suggestion
      <evan-form-control-input
```

File: dapps/assets/src/components/contacts/AddContact.vue [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+          v-model="idOrEmail"
+          :label="$t('_assets.contacts.id-or-email')"
```
Comment (Tschuck) [inline dapps/assets/src/components/contacts/AddContact.vue]: Also translation pipe can be used:
```suggestion
          :label="'_assets.contacts.id-or-email' | translate"
```

File: dapps/assets/src/components/contacts/Contacts.ts [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    this.data = await this.fetchContacts();    
+    this.isLoading = false;
+  }
+
+  handleRowClicked(contact: Contact) {
+    window.location.hash = `/${this.dapp.rootEns}/profile.vue.${this.dapp.domainName}/${contact.address}/detail`;
+  }
+
+  async handleContactAdded() {
+    console.log('Contact Added');
```
Comment (Tschuck) [inline dapps/assets/src/components/contacts/Contacts.ts]:
```suggestion
```

File: dapps/assets/src/components/contacts/Contacts.vue [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+
+<script lang="ts">
+import ContactsComponent from './Contacts';
+export default ContactsComponent;
+</script>
+
+<style lang="scss" scoped>
+@import '~@evan.network/ui/src/style/utils';
+
+.content {
```
Comment (Tschuck) [inline dapps/assets/src/components/contacts/Contacts.vue:130]: Already saw this one within the profile. We should add a core class for this.

File: dapps/assets/src/components/contacts/Contacts.vue [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+  color: cssVar('gray-600');
+  i.left.mdi {
+    margin-right: 0;
+  }
+  &.active {
+    border-bottom: 2px solid cssVar('primary');
+  }
+}
+
+/deep/ .evan-swipe-panel.light {
```
Comment (Tschuck) [inline dapps/assets/src/components/contacts/Contacts.vue:152]: Is this needed? Looks good for me on local machine. If yes, should be moved to the core!

File: dapps/assets/src/components/contacts/ContactsService.ts [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+    this.contacts = await this.runtime.profile.getAddressBook();
+
+    let data: Contact[] = [];
+    Object.keys(this.contacts.profile).forEach(async contact => {
+      const type = await this.getProfileType(contact);
+      data.push({
+        address: contact,
+        alias: this.contacts.profile[contact].alias,
+        createdAt: this.contacts.profile[contact].createdAt,
+        favorite: Math.random() > 0.5 ? true : false, // TODO: hard coded
```
Comment (Tschuck) [inline dapps/assets/src/components/contacts/ContactsService.ts:58]: Should be loaded and saved within the contact.

File: dapps/assets/src/components/contacts/InviteDispatcher.ts [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+  '_assets.contacts.invite-dispatcher'
+);
+
+/**
+ * Return a new bcc.Profile instance for a given accountId.
+ *
+ * @return     {bcc.Runtime}  profile for account
+ */
+const getProfileForAccount = (runtime: bcc.Runtime, accountId: string) => {
+  return new bcc.Profile({
```
Comment (Tschuck) [inline dapps/assets/src/components/contacts/InviteDispatcher.ts:37]: Could be also done by destructuring. But it think it was just taken from the old code. ;)

```
return new bcc.Profile({
  ...runtime,
  accountId
})
```

File: dapps/assets/src/components/digitaltwins/DigitalTwins.vue [hunk 13]
```diff
@@ -52,32 +56,50 @@ import DigitalTwinsComponent from './DigitalTwins';
 export default DigitalTwinsComponent;
 </script>
 
-<style lang="scss" scoped>
+<style lang="scss">
```
Comment (Tschuck) [inline dapps/assets/src/components/digitaltwins/DigitalTwins.vue:58]: We should never define any global styling within a component.  Please move evan-table to global ui-libs. Also `evan-table-body-row ` and `evan-table-head` can be removed. `evan-table` should be enough to style everything with one class.

```suggestion
<style lang="scss" scoped>
```

File: dapps/assets/src/components/Assets.ts [hunk 14]
```diff
@@ -30,6 +30,19 @@ declare module 'vue/types/vue' {
     $router: VueRouter;
     $route: Route;
     $store: any;
+    $t: Function;
+    dapp: Dapp;
+    getRuntime: Function;
+  }
+
+  interface Dapp {
```
Comment (karladler) [inline dapps/assets/src/components/Assets.ts:38]: should we add this to an interfaces file and just import it here? I guess it's commonly used.

File: dapps/assets/src/components/contacts/Contacts.vue [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+  <div>
+    <div class="content pt-5">
+      <div class="d-flex flex-row justify-content-between align-items-center">
+        <div>
+          <h1 class="h4" style="margin:0">{{ '_assets.contacts.contacts-title' | translate }}</h1>
+        </div>
+        <div>
+          <evan-button
+            @click="filterByFavorites()"
+            class="filter-btn ml-3"
```
Comment (Tschuck) [inline dapps/assets/src/components/contacts/Contacts.vue:30]: Should we add the .filter-btn class to the core, so it will be supported per default?

**Review by Tschuck — CHANGES_REQUESTED** [review]

**Comment by Tschuck** [comment]
Tldr. good work! Only some questions:

  - please add some comments to functions and variable definitions within components and classes
  - move table style to [ui-libs table](https://github.com/evannetwork/ui-core/blob/master/dapps/ui.libs/src/style/table.scss)
    - can we remove `evan-table-body-row` / `evan-table-head-row` classes?
  - update `nav-list` to work with router links
  - Should we add the .filter-btn class to the core, so it will be supported per default?
  - global `.content` style, with (max-width: 768px; margin-left:auto; margin-right:auto)

Some of the todos can also be done outside of the pr. We should only ensure, that we don't forget them. :)

File: dapps/assets/src/components/contacts/AddContact.ts (hunk 2 above)
Comment (karladler) [inline dapps/assets/src/components/contacts/AddContact.ts:54]: looks bit ugly because of the code duplications but I guess there is no other way around to have the vars reactive ...

File: dapps/assets/src/components/contacts/AddContact.ts [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+    }`;
+    this.msgTitle = this.$t('_assets.contacts.subject-prefill');
+  }
+
+  async addContact() {
+    if (this.checkFormValid()) {
+      const formData = {
+        accountId: this.accountId,
+        alias: this.alias,
+        createdAt: new Date().toISOString(),
```
Comment (karladler) [inline dapps/assets/src/components/contacts/AddContact.ts]: maybe create a dateString const first and apply it for both fields, it also ensures a 100% that created and updated are exactly the same.

File: dapps/assets/src/components/contacts/AddContact.ts [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+        currLang: window.localStorage.getItem('evan-language'),
```
Comment (karladler) [inline dapps/assets/src/components/contacts/AddContact.ts:79]: really need this????

File: dapps/assets/src/components/contacts/AddContact.ts [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+      this.accountId = null;
+      return '';
+    } else {
+      return '_assets.contacts.error-id-or-email';
+    }
+  }
+
+  checkFormValid(): boolean {
+    if (
+      document.querySelector('#contactForm :invalid') ||
```
Comment (karladler) [inline dapps/assets/src/components/contacts/AddContact.ts:128]: this looks hacky, I wouldn't rely on browser form validation to be honest.

File: dapps/assets/src/components/contacts/AddContact.vue (hunk 7 above)
Comment (karladler) [reply dapps/assets/src/components/contacts/AddContact.vue]: but I guess it's `:label="{'_assets.contacts.id-or-email' | translate"}` then ?

File: dapps/assets/src/components/Assets.ts (hunk 14 above)
Comment (datrinh) [reply dapps/assets/src/components/Assets.ts:38]: Normally this would be done in the component.ts and be available to all EvanComponents. Problem is currently modules arent exported properly so compiler doesn't understand. This should be fixed with TS-fixing instead.

File: dapps/assets/src/components/Assets.ts (hunk 1 above)
Comment (datrinh) [reply dapps/assets/src/components/Assets.ts:47]: This will work once we fix the TS-modules. Replay to Karl:

> Normally this would be done in the component.ts and be available to all EvanComponents. Problem is currently modules arent exported properly so compiler doesn't understand. This should be fixed with TS-fixing instead.

File: dapps/assets/src/components/contacts/AddContact.ts (hunk 2 above)
Comment (datrinh) [reply dapps/assets/src/components/contacts/AddContact.ts:54]: initState() is needed to reset the values when the sidepanel is closed. At least I didn't find a better (simple) solution to reset component state.

File: dapps/assets/src/components/contacts/Contacts.ts [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+  contactService: ContactsService;
+
+  isLoading = true;
+
+  sortBy = 'name';
+
+  filter = null;
+  filterBy: string[] = [];
+
+  data: Contact[] = [];
```
Comment (karladler) [inline dapps/assets/src/components/contacts/Contacts.ts]: take care with naming, since there is also this.$data in vue, not to get confused

File: dapps/assets/src/components/contacts/Contacts.ts [hunk 20]
```diff
@@ -0,0 +1,88 @@
...
+    this.data = await this.fetchContacts();
+    this.isLoading = false;
+  }
+
+  async fetchContacts(): Promise<Contact[]> {
+    return this.contactService.getContacts();
+  }
+
+  filterByType(type: string) {
+    this.filterBy = ['type'];
```
Comment (karladler) [inline dapps/assets/src/components/contacts/Contacts.ts:80]: what is the purpose of having `this.filterBy` as an array? Seems to hold always only on object?

File: dapps/assets/src/components/contacts/Contacts.vue [hunk 21]
```diff
@@ -0,0 +1,175 @@
...
+  Fifth Floor, Boston, MA, 02110-1301 USA, or download the license from
+  the following URL: https://evan.network/license/
+*/
+
+<template>
+  <div>
+    <div class="content pt-5">
+      <div class="d-flex flex-row justify-content-between align-items-center">
+        <div>
+          <h1 class="h4" style="margin:0">{{ '_assets.contacts.contacts-title' | translate }}</h1>
```
Comment (karladler) [inline dapps/assets/src/components/contacts/Contacts.vue:25]: `<h1 class="h4"` ❓

File: dapps/assets/src/components/contacts/Contacts.vue [hunk 22]
```diff
@@ -0,0 +1,175 @@
...
+          ></evan-button>
+          <evan-button
+            @click="filterByType('users')"
+            class="filter-btn ml-3"
+            type="text-secondary"
+            icon="mdi mdi-account-outline"
+            iconPosition="left"
+            :class="{ 'active': filter === 'users' }"
+            :label="$t('_assets.contacts.users')"
+          ></evan-button>
```
Comment (karladler) [inline dapps/assets/src/components/contacts/Contacts.vue]: I think this can all be written as self closing component.

`/>` instead of `></evan-button>`

File: dapps/assets/src/components/contacts/Contacts.vue [hunk 23]
```diff
@@ -0,0 +1,175 @@
...
+          :tbody-tr-class="'evan-table-body-row'"
+          :thead-tr-class="'evan-table-head-row'"
+          :thead-class="'evan-table-head'"
+          @row-clicked="handleRowClicked"
+          sticky-header="80vh"
+          show-empty
+        >
+          <template
+            v-slot:cell(alias)="data"
+          >{{ data.item.alias ? data.item.alias: data.item.address }}</template>
```
Comment (karladler) [inline dapps/assets/src/components/contacts/Contacts.vue]:
```suggestion
          >{{ data.item.alias ? data.item.alias : data.item.address }}</template>
```

File: dapps/assets/src/components/contacts/Contacts.vue (hunk 10 above)
Comment (karladler) [reply dapps/assets/src/components/contacts/Contacts.vue:152]: if needed should be added to the swipe panel. Maybe a theme mixin.

File: dapps/assets/src/components/contacts/Contacts.vue [hunk 24]
```diff
@@ -0,0 +1,175 @@
...
+  &.active {
+    border-bottom: 2px solid cssVar('primary');
+  }
+}
+
+/deep/ .evan-swipe-panel.light {
+  .light & {
+    background-color: cssVar('body-bg');
+
+    &,
```
Comment (karladler) [inline dapps/assets/src/components/contacts/Contacts.vue]: I would put it in one line. What does csscomb here?

File: dapps/assets/src/components/contacts/AddContact.ts (hunk 2 above)
Comment (datrinh) [reply dapps/assets/src/components/contacts/AddContact.ts:54]: Nope, there isn't a better solution, unless we hack some different logic into it to reset component state dynamically.

File: dapps/assets/src/components/contacts/Contacts.vue (hunk 9 above)
Comment (karladler) [reply dapps/assets/src/components/contacts/Contacts.vue:130]: also I'm not happy with inline stylings, would it be possible to commit to put stylings always in *scss files?

File: dapps/assets/src/components/contacts/Contacts.vue [hunk 25]
```diff
@@ -0,0 +1,175 @@
...
+<script lang="ts">
+import ContactsComponent from './Contacts';
+export default ContactsComponent;
+</script>
+
+<style lang="scss" scoped>
+@import '~@evan.network/ui/src/style/utils';
+
+.content {
+  max-width: 768px;
```
Comment (karladler) [inline dapps/assets/src/components/contacts/Contacts.vue:131]: think this is cssvar breakpoint sm or something

File: dapps/assets/src/components/contacts/AddContact.ts (hunk 17 above)
Comment (datrinh) [reply dapps/assets/src/components/contacts/AddContact.ts:79]: You mean currLang? I'm not sure. I took the logic from existing addressbook tbh.

File: dapps/assets/src/components/contacts/AddContact.ts (hunk 18 above)
Comment (datrinh) [reply dapps/assets/src/components/contacts/AddContact.ts]: Since it only checks required status, I think this will work very reliably. I'm more confident in native browser checks than custom js validation tbh :)

File: dapps/assets/src/components/contacts/Contacts.ts (hunk 20 above)
Comment (datrinh) [reply dapps/assets/src/components/contacts/Contacts.ts:80]: It's what bootstrap-table expects. Potentially you could filter by multiple fields (which we don't do in our case)

File: dapps/assets/src/components/contacts/Contacts.vue (hunk 21 above)
Comment (datrinh) [reply dapps/assets/src/components/contacts/Contacts.vue:25]: h1 for semantic html and h4 for styling ;)

File: dapps/assets/src/components/contacts/Contacts.vue (hunk 25 above)
Comment (datrinh) [reply dapps/assets/src/components/contacts/Contacts.vue:131]: We aren't really using any breakpoint vars anywhere from what Daniel said and I saw so far. I'm also not sure if the content max-width will always correlate with some breakpoint.

File: dapps/assets/src/components/contacts/AddContact.ts (hunk 18 above)
Comment (datrinh) [reply dapps/assets/src/components/contacts/AddContact.ts:128]: Since it only checks required status, I think this will work very reliably. I'm more confident in native browser checks than custom js validation tbh :)

File: dapps/assets/src/components/contacts/ContactsService.ts (hunk 11 above)
Comment (datrinh) [reply dapps/assets/src/components/contacts/ContactsService.ts:58]: What exactly do you mean by that?

File: dapps/assets/src/components/contacts/AddContact.vue (hunk 7 above)
Comment (datrinh) [reply dapps/assets/src/components/contacts/AddContact.vue]: Yes, translate pipe would work. But I think it's longer :p
If we agree on a certain style for the templates, we should do this during xmas clean up :)

File: dapps/assets/src/components/contacts/Contacts.vue (hunk 9 above)
Comment (datrinh) [reply dapps/assets/src/components/contacts/Contacts.vue:130]: We should discuss this in general, because I'm actually a fan of keeping all template related things in one file and not pollute the folder structure with so many partials. Should do this during retro or HQOTD

File: dapps/assets/src/components/digitaltwins/DigitalTwins.vue (hunk 13 above)
Comment (datrinh) [reply dapps/assets/src/components/digitaltwins/DigitalTwins.vue:58]: Yes, this is a quick solution to share the evan-table style for both tables. This will be done in a refactor ticket (https://evannetwork.atlassian.net/browse/CORE-791)

File: dapps/assets/src/components/contacts/Contacts.vue (hunk 15 above)
Comment (datrinh) [reply dapps/assets/src/components/contacts/Contacts.vue:30]: See here: https://evannetwork.atlassian.net/browse/CORE-793

File: dapps/assets/src/components/contacts/InviteDispatcher.ts (hunk 12 above)
Comment (datrinh) [reply dapps/assets/src/components/contacts/InviteDispatcher.ts:37]: Your suggested code throws errors. I'm not sure how to do it like you intend it to.

File: dapps/assets/src/components/contacts/Contacts.vue (hunk 10 above)
Comment (datrinh) [reply dapps/assets/src/components/contacts/Contacts.vue:152]: Added it here: https://evannetwork.atlassian.net/browse/CORE-794

**Comment by datrinh** [comment]
I collected and extracted the needed refactoring for the next (xmas) sprint:
https://evannetwork.atlassian.net/browse/CORE-792

File: dapps/assets/src/i18n/de.json [hunk 26]
```diff
@@ -1,12 +1,45 @@
 {
   "_assets": {
-    "contacts": "Kontakte",
-    "digitaltwins": "Digitale Zwillinge",
-    "digitaltwin": {
-      "digitaltwin-title": "Digitale Zwillinge",
-      "my-own": "Eigene",
+    "contacts": {
+      "add-contact-btn": "Kontakt hinzufügen",
+      "add-contact-desc": "Geben Sie die Evan-ID oder die E-Mail-Adresse Ihres Kontaktes ein um diesen einzuladen.",
```
Comment (karladler) [inline dapps/assets/src/i18n/de.json]: Das wording ist hier etwas ungeschickt, ich denke es ist nicht ganz klar, dass man entweder die Evan ID eines bestehenden Nutzers nehmen sollte, oder die email um einen Nutzer neu einzuladen. Email für bestehenden Nutzer geht ja nicht.

File: dapps/assets/src/components/contacts/InviteDispatcher.ts (hunk 12 above)
Comment (datrinh) [reply dapps/assets/src/components/contacts/InviteDispatcher.ts:37]: Ah, I see. This would need an update on the bcc.Runtime interface. With typing, it throws an error.

File: dapps/assets/src/components/contacts/AddContact.ts (hunk 2 above)
Comment (Tschuck) [reply dapps/assets/src/components/contacts/AddContact.ts:54]: Ah, i see. then just set all of the head variables to `null`, so we avoid duplicated code and to make it reative.

File: dapps/assets/src/components/contacts/AddContact.vue (hunk 7 above)
Comment (Tschuck) [reply dapps/assets/src/components/contacts/AddContact.vue]: i'm fine with that, just wanted to mention. ;)

File: dapps/assets/src/components/contacts/ContactsService.ts (hunk 11 above)
Comment (Tschuck) [reply dapps/assets/src/components/contacts/ContactsService.ts:58]: I mean, we should implement the save and loading logic for the favorite flag for the contacts^^

File: dapps/assets/src/components/contacts/AddContact.ts (hunk 2 above)
Comment (Tschuck) [reply dapps/assets/src/components/contacts/AddContact.ts:54]: As i mentioned before, we can just set initial values to null, so we avoid duplicated code, but keep reactivity

File: dapps/assets/src/components/contacts/AddContact.ts (hunk 17 above)
Comment (Tschuck) [reply dapps/assets/src/components/contacts/AddContact.ts:79]: It's required for the blockchain-core invitation:

```
await runtime.onboarding.sendInvitation(
  {
    to: accountId,
    subject: data.msgTitle,
    body: data.msgBody,
    fromAlias: data.fromAlias,
    lang: data.currLang
  },
  runtime.web3.utils.toWei('0')
);
```

File: dapps/assets/src/components/contacts/InviteDispatcher.ts (hunk 12 above)
Comment (Tschuck) [reply dapps/assets/src/components/contacts/InviteDispatcher.ts:37]: Was just a quick recommendation. We should not change bcc.Runtime interface for that. ProfileOptions are different from the runtime, but basically just a reduced runtime, including the `defaultCryptoAlgo` param. We can just force the runtime to be applied to the profile options.

File: dapps/assets/src/components/contacts/AddContact.ts (hunk 18 above)
Comment (Tschuck) [reply dapps/assets/src/components/contacts/AddContact.ts:128]: Would also suggest the same.
Hint again: We can also use, the `EvanForm` declaration without automatic formular generation. Just use the `EvanForm` as `value` and validation "holder". Code is not tested, only copied from somewhere :D 

```ts
this.form = (<IssueFormInterface>new EvanForm(this, {
  idOrEmail: {
    value: '',
    validate: function(vueInstance: IssueComponent, form: IssueFormInterface) {
      if (EvanForm.validEthAddress(this.value)) {
        if (vueInstance.addressbook.profile[this.value]) {
          return '_assets.contacts.error-added';
        } else {
          this.emailInvite = false;
          vueInstance.accountId = value;
          vueInstance.email = null;
          return '';
        }
      } else if (EvanForm.validateEmail(this.value)) {
        vueInstance.emailInvite = true;
        vueInstance.email = value;
        vueInstance.accountId = null;
        return '';
      } else {
        return '_assets.contacts.error-id-or-email';
      }
    }
  },
```

```html
<evan-form-control-input
  v-model="form.idOrEmail.value"
  :label="'_assets.contacts.id-or-email' | translate"
  :placeholder="$t('_assets.contacts.id-or-email-placeholder')"
  :required="true"
  :error="form.idOrEmail.error"
  @input="handleIdOrEmailChange"
/>

...

<evan-button
  type="primary"
  native-type="submit"
  class="ml-3 flex-grow-1"
  :disabled="form.invalid"
  :label="'_assets.contacts.add-contact-btn' | translate"
/>
```

File: dapps/assets/src/i18n/de.json (hunk 26 above)
Comment (datrinh) [reply dapps/assets/src/i18n/de.json]: Yeah, I find the concept of mixing inviting existing and new accounts troublesome too. But I just followed the wireframes for now.

File: dapps/assets/src/components/contacts/AddContact.ts (hunk 18 above)
Comment (datrinh) [reply dapps/assets/src/components/contacts/AddContact.ts:128]: This would require refactoring the other inputs into evan form too... I tried it and stashed it for now since it requires more refactoring than I thought :D

Also it's not "hacky" or anything. It's well documented here: https://developer.mozilla.org/en-US/docs/Learn/HTML/Forms/Form_validation
It's a good solution for simple validation like "required".

**Review by Tschuck — APPROVED** [review]

File: dapps/assets/src/i18n/de.json [hunk 27]
```diff
@@ -1,12 +1,46 @@
...
+      "add-contact-title": "Neuen Kontakt hinzufügen",
+      "all": "Alle",
+      "cancel": "Abbrechen",
+      "companies": "Unternehmen",
+      "contact-name-placeholder": "Max Mustermann",
+      "contact-name": "Kontaktname",
+      "contacts-empty": "Sie haben noch keine Kontakte hinzugefügt.",
+      "contacts-title": "Kontakte",
+      "created": "Erstellt",
+      "email-desc": "Der ausgewählte Nutzer wird mit Ihrer Anfrage kontaktiert. Im folgenden Formular können Sie diese bearbeiten.",
```
Comment (wulfraem) [inline dapps/assets/src/i18n/de.json:15]: actually no idea what the text should be, but this one sounds a bit strange... or? oO

Comment (datrinh) [reply dapps/assets/src/i18n/de.json:15]: Im open for suggestions :D
